pkg/cmd/pr/create/create.go

```diff
@@ -72,16 +73,93 @@ type CreateOptions struct {
 	DryRun bool
 }
 
+// refs is an interface that provides the necessary information for creating a pull request in the API.
+// Upcasting to concrete implementations can provide further context on other operations (forking and pushing).
+type refs interface {
+	QualifiedHeadRef() string
+	UnqualifiedHeadRef() string
+	BaseRef() string
+
+	// While the only thing really required from an api.Repository is the repository ID, changing that
+	// would require changing the API function signatures, and the refactor that introduced this refs
+	// type is already large enough.
+	BaseRepo() *api.Repository
+}
+
+type baseRefs struct {
+	baseRepo       *api.Repository
+	baseBranchName string
+}
+
+func (r baseRefs) BaseRef() string {
+	return r.baseBranchName
+}
+
+func (r baseRefs) BaseRepo() *api.Repository {
+	return r.baseRepo
+}
+
+type skipPushRefs struct {
+	qualifiedHeadRef shared.QualifiedHeadRef
+
+	baseRefs
+}
+
+func (r skipPushRefs) QualifiedHeadRef() string {
+	return r.qualifiedHeadRef.String()
+}
+
+func (r skipPushRefs) UnqualifiedHeadRef() string {
+	return r.qualifiedHeadRef.BranchName()
+}
+
+type pushableRefs struct {
+	headRepo       ghrepo.Interface
+	headBranchName string
+
+	baseRefs
+}
+
+func (r pushableRefs) QualifiedHeadRef() string {
+	if ghrepo.IsSame(r.headRepo, r.baseRepo) {
+		return r.headBranchName
+	}
+	return fmt.Sprintf("%s:%s", r.headRepo.RepoOwner(), r.headBranchName)
+}
+
+func (r pushableRefs) UnqualifiedHeadRef() string {
+	return r.headBranchName
+}
+
+func (r pushableRefs) HeadRepo() ghrepo.Interface {
+	return r.headRepo
+}
+
+// TODO: consider not embedding here, because it's a bit misleading
+type forkableRefs struct {
+	qualifiedHeadRef shared.QualifiedHeadRef
+
+	baseRefs
+}
+
+func (r forkableRefs) QualifiedHeadRef() string {
+	return r.qualifiedHeadRef.String()
+}
+
+func (r forkableRefs) UnqualifiedHeadRef() string {
+	return r.qualifiedHeadRef.BranchName()
+}
```

Comment:
Suggestion: would you be open to a comment here?

```suggestion
// QualifiedHeadRef returns a head ref in the form of `<user>:<branch>`. 
// This is typically used when interacting with the GitHub API to disambiguate
// the head repository in pull request scenarios involving forks.
func (r forkableRefs) QualifiedHeadRef() string {
	return r.qualifiedHeadRef.String()
}

// UnqualifiedHeadRef returns a head ref in the form of the branch name only.
func (r forkableRefs) UnqualifiedHeadRef() string {
	return r.qualifiedHeadRef.BranchName()
}
```